### DISCUSSION TITLE
Incoming emails that lie about their Content-Type

### DISCUSSION THREAD

**Discussion by dcpurton** [neomutt/neomutt#3231] [Development]
It's common for automated mailers to be quite broken in how they produce emails.

Consider a PGP encrypted email from Facebook (once decrypted):

```
  I     1 <no description>                                              [multipart/alternative, 7bit, 24K]
  I     2 └─><no description>                                               [text/plain, 7bit, utf-8, 23K]
```

Leaving aside the fact that there's a `multipart/alternative` MIME part with only one child (*sigh*), the one part it does contain is marked as `text/plain`.

Except it's not. In reality the attachment is `text/html`. So Neomutt just renders out the HTML as plain text.

I don't read machine produced HTML so quickly, so it would be good to have some kind of way of catching this situation and letting Neomutt know that the attachment is `text/html`.

At some point in the past (before the last release). It was possible to change the MIME type of attachments in incoming emails as a kind of workaround. So I used to select the attachment and `<edit-type>` to change it to `text/html`.

But if I try and do this now, I get an error:

```
Structural changes to decrypted attachments are not supported
```

Obviously this sounds reasonable, but removes the ability to easily see what these emails say. Worse, of the mail readers I use only Neomutt supports PGP, so I can't look at them in something else either. The only way to see what Facebook has sent is to save the attachment as `something.html` and then open that file in a browser.

9 times out of 10, Facebook just sends me an ad trying to get me to spend money to promote something, but occasionally it's something relevant.

Looking for thoughts on how to deal with this situation. Ideally with an automated hook that enables me to transparently just read these broken emails.

I have complained to Facebook about this, but we all know what will happen with that kind of complaint. If only commercial software companies had support as good as you see in the Open Source world…

If it's useful, I can produce a bisect that shows where the old behaviour stopped working.

**Comment by flatcap** [comment]
> `Structural changes to decrypted attachments are not supported`

I looked up the error message.  It dates back over 4 years (an upstream commit).
I _think_ I can understand the reason for not making **structural** changes,
but temporarily altering the mime type shouldn't be a problem.

---

My first thought was the `mime_lookup` command, but that seems to require a filename.

Next I tried a custom `mailcap` file that treated `text/plain` to `links`.  No joy.

I re-encoded my email with:
- Fake mime type of `text/wibble`
- Configured `auto_view text/wibble`
- Mailcap: `text/wibble; links -dump %s; copiousoutput`

_That_ worked, which implies that `text/plain` is being treated differently: `mailcap` isn't being used.

---

Should we allow temporary cosmetic changes to decrypted emails?

Should we always honour the `mailcap`, even for `text/plain`?

**Reply by dcpurton** [reply]
Yeah, I saw that it had been in for a while. Which is odd. Because I have viewed html parts like this in the last 12 months with the main version of Neomutt in Debian.

**Comment by dcpurton** [comment]
Bisect finds:

459957955d4c274febb6cbc6b3dd24d9b7a33006 is the first bad commit
commit 459957955d4c274febb6cbc6b3dd24d9b7a33006
Author: Marco Sirabella <marco@sirabella.org>
Date:   Wed Mar 31 21:19:48 2021 -0700

Return type_changes too for mutt_edit_content_type
    
This fixes #2875
    
The only other place this function is used other than to calculate when
to redraw changes in compose is in `recvattach_edit_content_type`, but I
think this change improves that.
    
Not exactly sure of the context around why a user would want to change
the content type of a recieved email, but I can't trigger the longer
path in that function without the changes in this commit
    
It's either this or undoing the redraw check made in 49e4decd88
    
commands.c | 4 ++--
1 file changed, 2 insertions(+), 2 deletions(-)

**Comment by flatcap** [comment]
Workaround:

**mailcap**
```
text/plain; links -dump %s; copiousoutput
```

**neomuttrc**
```
set mailcap_path="mailcap"
```

**NeoMutt**
- Index: `<view-attachments>`
- Attach: `<view-mailcap>`

That forcibly uses the mailcap entry for `text/plain`.

**Comment by dcpurton** [comment]
I just realised I can do this with hooks too:

```ini
auto_view  text/html
message-hook . 'unauto_view  text/plain'
message-hook '~f facebookmail.com' 'auto_view text/plain'
```
